src/metabase/lib/query.cljc

```diff
@@ -387,3 +388,36 @@
     (-> a-query
         (update :stages #(vec (take (inc stage-number) %)))
         (update-in [:stages stage-number] preview-stage clause-type clause-index))))
+
+(mu/defn wrap-native-query-with-mbql :- [:map
+                                         [:query ::lib.schema/query]
+                                         [:stage-number :int]]
+  "Given a query and stage number, return a possibly-updated query and stage number which is guaranteed to be MBQL and
+  so to support drill-thru and similar logic. Such a query must be saved, hence the `card-id`.
+
+  If the provided query is already MBQL, this is transparent.
+
+  Returns `{:query query', :stage-number stage-number'}`.
+
+  You might find it more convenient to call [[with-wrapped-native-query]]."
+  [a-query      :- ::lib.schema/query
+   stage-number :- :int
+   card-id      :- [:maybe ::lib.schema.id/card]]
+  (or (and (lib.util/native-stage? a-query stage-number)
+           card-id
+           (if-let [card (lib.metadata/card a-query card-id)]
+             {:query        (query a-query card)
+              :stage-number -1}
+             (do
+               (log/warn "Failed to wrap native query with MBQL; card not found" {:query   a-query
+                                                                                  :card-id card-id})
+               nil)))
+      {:query        a-query
+       :stage-number stage-number}))
```

Comment:
Nit: docstring states returned is guaranteed to be mbql. I think that's not the case with native and failed card lookup. Or is it?